azkaban-common/src/main/java/azkaban/project/ProjectLoader.java

```diff
@@ -161,6 +162,37 @@ List<Flow> fetchAllProjectFlows(final Project project)
   Map<Project, List<Flow>> fetchAllFlowsForProjects(List<Project> projects)
       throws ProjectManagerException;
 
+  /**
+   * Should create an empty flow recommendation with the given projectId and flowId and adds it to
+   * the data store. It will auto assign a unique id for this flow recommendation if successful.
+   */
+  FlowRecommendation createFlowRecommendation(int projectId, String flowId)
+      throws ProjectManagerException;
+
+  /**
+   * Update one particular flow recommendation
+   */
+  void updateFlowRecommendation(FlowRecommendation flowRecommendation)
+      throws ProjectManagerException;
+
+  /**
+   * Fetches one particular flow recommendation.
+   */
+  FlowRecommendation fetchFlowRecommendation(int projectId, String flowId)
+      throws ProjectManagerException;
+
+  /**
+   * Fetches all flow recommendations for a given project
+   */
+  List<FlowRecommendation> fetchAllProjectFlowRecommendations(Project project)
```

Comment:
Also I feel like passing through the entire Project all the way here could have the chance to increase the Java garbage collection burden so it might result in frequent gc to our fragile web server, as Project itslef contains too many informations. (Not sure about the numbers but I've already feel like our current preparing/loading process is too long)
If the purpose is only a project identifier, can we use project_id only perhaps?

Reply:
1. "Passing entire Project" is coming from the existing fetchAllFlowsForProjects logic and creating one more list of projectIds in addition to existing list of projects in InMemoryProjectCache seems an overkill.
2. "Our current preparing/loading process is too long": currently it is b/c we have too many projects and do not get retentioned/deleted since most of them are really old. I don't think using projectIds here will reduce load speed a lot b/c projectIds should be generated in outer level as well from projects. It is the same effort: difference is where to get id from project: in outer class or ProjectLoader class but it will not affect performance a lot.

Reply:
Agree with extra mapping is an overkill. My opinion is from garbage collection point of view. I used to work on a project, we pass in a complex object into multiple layer and this is on critical path, the memory usage increased exponentially and the service crashed frequently since then. Because in each method of that call tree, such huge object would occupy a lot memory (java is copy-by-value instead of copy-by-reference) and such method is being called so frequently. It then ends up triggering heavy JGC. I am sensitive to see such and would think ahead if we really need whole object to pass through.
Anyway I agree it might not be a big deal now. However, when the projects keep growing, the headache would become more serious as I think we're already in a situation to clean up unused projects sometime or so. So what is already existing, might not be a best way to follow. Let always keep in critical mind for things: )

Reply:
After offline discussion, copy-by-value in java does not mean deep copy the object but means copy the object pointer/reference for non-primitive objects so it is lightweight no matter how the huge the object itself is.
https://www.infoworld.com/article/3512039/does-java-pass-by-reference-or-pass-by-value.html